File: product-service/handlers.js
const { publishEvent } = require("./pubsub");
const Product = require('./models/products.model');

const processedMessages = new Set();

async function handleCartEvent(message) {
    const { event, data } = message;
    console.log("Handling cart event");
    let productsList = [];
    if (event === 'cart:ended') {
        try {
            let total = 0;
            for (let p of data) {
                const product = await Product.findByPk(p.product_id);
                console.log(product.dataValues);
                if (product.dataValues.stock < p.quantity) {
                    console.log(`Product ${product.dataValues.name} is out of stock`);
                    publishEvent('cart:cancelled', 'cart', { cart_id: data.cart_id });
                    return;
                }
                productsList.push({
                    id: product.dataValues.id,
                    name: product.dataValues.name,
                    quantity: p.quantity,
                    unitaryPrice: product.dataValues.price,
                    totalPrice: product.dataValues.price * p.quantity
                });
                const newStock = product.dataValues.stock - p.quantity;
                await Product.update({ stock: newStock }, {
                    where: { id: product.id }
                });
                console.log(`Stock updated for product ${product.dataValues.name}`);
                total += product.dataValues.price * p.quantity;
            }
            console.log(productsList);
            console.log(`Total amount charged: ${total}`);
            publishEvent('cart:processed', 'cart', { cart_id: data[0].cart_id, amount: total, productsList: productsList });

        } catch (error) {
            console.log(error.message);
        }
    }
}

async function handleProductEvent(message) {
    const { event, data, message_id } = message;
    if (!message_id) {
        console.log("Message ID is missing");
        return;
    }
    if (event === 'products:decrementInventory') {
        try {
            console.log(data.productList);
            let productsList = JSON.parse(data.productList);
            console.log(productsList);
            for (let p of productsList) {
                console.log(p.id);
                const product = await Product.findByPk(p.id);
                if (!product) {
                    console.log(`Product with id ${p.id} not found`);
                    continue;
                }
                const newStock = product.dataValues.stock - p.quantity;
                await Product.update({ stock: newStock }, {
                    where: { id: product.id }
                });
                console.log(`Stock updated for product ${product.dataValues.name}`);
            }
        } catch (error) {
            console.log(error.message);
        }
    }
}

module.exports = {
    handleCartEvent,
    handleProductEvent
};

File: cart-service/handlers.py
import json
from sqlalchemy.orm import Session
from fastapi import HTTPException
from models.users import Users
from models.products import Products
from fastapi.encoders import jsonable_encoder
import datetime
from database.database import SessionLocal
import asyncio
from models.eventlog import EventLog
processed_messages = set()

def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()

def handle_message(message):
    from pubsub import publish_event
    session = next(get_db())
    if message and message["type"] == "message":
        payload = json.loads(message["data"])
        event = payload["event"]
        data = payload["data"]
        message_id = payload.get('message_id')

        print(f"Message ID: {message_id}")
        if message_id in processed_messages:
            print(f"Message {message_id} already processed")
            return
        processed_messages.add(message_id)
        print("Mensajes procesados", processed_messages)

        if session.query(EventLog).filter(EventLog.event_id == message_id).first():
            print("Evento ya procesado anteriormente")
            return 
        event_log = EventLog(
                event_id=message_id,
                event_type=event,
                processed_at=datetime.datetime.now()
            )
        session.add(event_log)
        if event == 'product:created':
            print(f"Product added: {data}")
            if session.query(Products).filter(Products.product_id == data['product_id']).first():
                raise HTTPException(status_code=400, detail="Product already exists")
            product = Products(product_id=data["product_id"], created_at=datetime.datetime.now(), updated_at=datetime.datetime.now())
            session.add(product)
            session.commit()
            session.refresh(product)
            return product
        if event == 'product:deleted':
            print(f"Product deleted: {data['product_id']}")
            product = session.query(Products).filter(Products.product_id == data['product_id']).first()
            if product:
                session.delete(product)
                session.commit()
                return product
            else:
                raise HTTPException(status_code=404, detail="Product not found")
        if event == 'user:created':
            print(f"User created: {data['user_id']}")
            if session.query(Users).filter(Users.user_id == data['user_id']).first():
                print("Este usuario existe ")
                return
            user = Users(
                user_id=data["user_id"],
                created_at=datetime.datetime.now(),
                updated_at=datetime.datetime.now()
            )            
            session.add(user)
            session.commit()
            session.refresh(user)
            print("Final")
            return user

        if event == 'user:deleted':
            print(f"User deleted: {data['user_id']}")
            user = session.query(Users).filter(Users.user_id == data['user_id']).first()
            if user:
                session.delete(user)
                session.commit()
                return user
            else:
                raise HTTPException(status_code=404, detail="User not found")
        if event == 'cart:processed':
            dataToSent = {
                "cart_id": data['cart_id'],
                "amount": data['amount'],
                "productList": data['productsList']
            }
            print("Data to send", dataToSent)
            publish_event("payment", "payment:payment_send", dataToSent)
            return dataToSent